fraud/service.go

```diff
@@ -0,0 +1,132 @@
+package fraud
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"sync"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+// service is propagating and validating Fraud Proofs
+// service implements Service interface.
+type service struct {
+	topics *topics
+
+	unmarshalers map[ProofType]ProofUnmarshaler
+	mu           sync.RWMutex
+
+	getter headerFetcher
+}
+
+func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
+	return &service{
+		topics: &topics{
+			pubsub:       p,
+			pubSubTopics: make(map[ProofType]*pubsub.Topic),
+		},
+		unmarshalers: make(map[ProofType]ProofUnmarshaler),
+		getter:       getter,
+	}
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.mu.RLock()
+	u, ok := f.unmarshalers[proofType]
+	f.mu.RUnlock()
+	if !ok {
+		return nil, errors.New("fraud: unmarshaler is not registered")
+	}
+
+	t, wasNotjoined, err := f.topics.getTopic(proofType)
+	if err != nil {
+		return nil, err
+	}
+	// if topic was joined for the first time then we should register a validator for it.
+	if wasNotjoined {
```

Comment:
This concept is very confusing. If a topic gets added to `topics`, it must also be required that a validator is mounted onto the topic before it is registered on topics. The current design is quite messy.

Reply:
I tend to agree here. Not sure how to make it better here besides integrating topics into the service.

Reply:
I can suggest :

```
func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
	f.mu.Lock()
	u, ok := f.unmarshalers[proofType]
	
	if !ok {
		return nil, errors.New("fraud: unmarshaler is not registered")
	}

       // returns only cached topic
	t, ok := f.topics.getTopic(proofType)
	if !ok {
		if err = f.topics.registerValidator(proofType, f.processIncoming); err != nil {
                        f.mu.Unlock()
			return nil, err
		}
                // register topic
	}
       f.mu.Unlock()
	return newSubscription(t, u)
}
```